dev_guide/builds.adoc

```diff
@@ -549,189 +615,489 @@ source:
+      },
+      "type": "Git"
+    },
+    "strategy": {
+      "sourceStrategy": {
+        "from": {
+          "kind": "ImageStreamTag",
+          "name": "python-33-centos7:latest"
+        }
+      },
+      "type": "Source"
+    }
+  }
+----
+<1> The URL of private repository, accessed by a private SSH key, is usually
+in the form `git@example.com:<username>/<repository>.git`.
+====
+
+[[other-authentication]]
+==== Other
+
+In case the cloning of your application is dependent on a `CA certificate`, `.gitconfig`
+file or both, you can create a secret that contains them, add it to the builder service
+account and then your `BuildConfig`.
+
+. Create desired type of `*secret*`:
+
+.. To create a secret from a `.gitconfig`:
++
+====
+----
+$ oc secrets new mysecret .gitconfig=path/to/.gitconfig
+----
+====
+.. To create a secret from a `CA certificate`:
++
+====
+----
+$ oc secrets new mysecret ca.crt=path/to/certificate
+----
+====
+.. To create a secret from a `CA certificate` and `.gitconfig`:
++
+====
+----
+$ oc secrets new mysecret ca.crt=path/to/certificate .gitconfig=path/to/.gitconfig
+----
+====
+
+[NOTE]
+====
+Please note that SSL verification can be turned off, if `sslVerify=false` is set
+for the `http` section in your `.gitconfig` file:
+----
+[http]
+        sslVerify=false
+----
+====
+
+. Add the `*secret*` to the builder service account:
++
+====
+----
+$ oc secrets add serviceaccount/builder secrets/mysecret
+----
+====
+
+
+[[dockerfile-source]]
+
+== Dockerfile Source
+
+When the `*BuildConfig.spec.source.type*` is `*Dockerfile*`, an inline
+Dockerfile is used as the build input, and no additional sources can be
+provided.
+
+This source type is valid when the build strategy type is `*Docker*` or
+`*Custom*`.
+
+The source definition is part of the *spec*` section in the `*BuildConfig*`:
+
+====
+
+----
+{
+ "source" : {
+    "type" : "Dockerfile",
+    "dockerfile": "FROM centos:7\nRUN yum install -y httpd" <1>
+ },
+}
+----
+
+<1> The `*dockerfile*` field contains an inline Dockerfile that will be built.
+====
+
+
+[[binary-source]]
+
+== Binary Source
+
+When the `*BuildConfig.spec.source.type*` is `*Binary*`, the build will expect a
+binary as input, and an inline Dockerfile is optional.
+
+The binary is generally assumed to be a tar, gzipped tar, or zip file depending
+on the strategy. For `*Docker*` builds, this is the build context and an
+optional Dockerfile may be specified to override any Dockerfile in the build
+context. For `*Source*` builds, this is assumed to be an archive as described
+above. For `*Source*` and `*Docker*` builds, if `*binary.asFile*` is set the
+build will receive a directory with a single file. `*contextDir*` may be used
+when an archive is provided. Custom builds will receive this binary as input on
+standand input (`stdin`).
+
+A binary source potentially extracts content, in which case `*contextDir*`
+allows changing to a subdirectory within the content before the build executes.
```

Comment:
Some questions i'd have as a reader:
- Is there any check that `contextDir` is a relative filename (such that it forms a valid subdirectory)?
- Can it be an arbitrary or absolute directory name (e.g., `/tmp/spacious`)?
- Whatever the name, does that directory need to be present or will the build process create it if it does not yet exist?